### DISCUSSION TITLE
How can I get all users?

### DISCUSSION THREAD

**Discussion by amr3k** [fastapi-users/fastapi-users#520] [Q&A] [answered]
Hello, thanks for this great module, it saved me a lot of time.
I'm wondering if it is possible for a super user to view all users in a single request?
Something like `GET /users` which returns
```
{
    "user1" : { ....data.... },
    "user2" : { ....data.... },
    "user3" : { ....data.... },
}
```
I'd be great if it's possible to use filters as well like `GET /users?is_active=1` and so on.

Am I asking in the wrong place? should I use a different module (like motor or odmantic) to handle database queries? and if that's the case how do I use fastapi_users with such module?

Thanks again.

**Comment by aaronamelgar** [comment]
You'll want to specify that route in fastapi with a superuser dependency and then write a custom query the user table, returning a list of all users. If you are using SQLAlchemyUserDatabase as your ORM you can just use a standard .all() query method.

**Reply by amr3k** [reply]
I'm using mongo but haven't decided yet which library to use :/ 
Could I make a class which inherits `fastapi_users.db.MongoDBUserDatabase` and add a method that get all users ?

**Comment by frankie567** [comment, marked as answer]
Hey there 👋 You are in the right place to ask questions, thanks 😄

In the first versions, there used to be such endpoint. I removed it because it can vary greatly from one project to another : how you handle pagination ? Which users can call the endpoint? Which users can you see according to your permissions? And so on.

So I prefer the developers to implement it themselves so it perfectly fits their need 😃 It's quite easy to do, it's just standard FastAPI and will look like a lot your own endpoints for your own data. Here is an example assuming you're using MongoDB:

```py
@app.get("/users")
async def list_users(
  is_active: bool = Query(None),  # Optional query parameter "is_active"
):
  users_collection = db["users"]  # `db` is an instance of `AsyncIOMotorDatabase` (just like in the FastAPI Users exampleà
  query: Dict[str, Any] = {}  # Start to build a query (empty query means everything in MongoDB)

  # Apply an is_active filter if the query is specified
  if is_active is not None:
    query["is_active"] = is_active

  # Perform the query
  cursor = users_collection.find(query)  # This an async iterator
  results = [User(**obj) async for obj in cursor]  # For each result, MongoDB gives us a raw dictionary that we hydrate back in our Pydantic model

  return results
```

Up to you after to do more clever things with other filters, pagination, etc. Basically, it's just building MongoDB queries.

**Reply by AndreWohnsland** [reply]
Hey, @frankie567 would it also be possible to provide a code example with the needed imports for the default [SQLAlchemy](https://fastapi-users.github.io/fastapi-users/11.0/configuration/full-example/#sqlalchemy) complete example provided by the documentation? I am currently trying to implement a custom route which returns all existing users in the SQL database, but having problems since I have little working experience with SQLAlchemy.
